dataset/bins.cpp

```diff
@@ -405,4 +405,36 @@ Dataset sum(const Dataset &d) {
   return apply_to_items(d, [](auto &&... _) { return buckets::sum(_...); });
 }
 
+Variable mean(const Variable &data) {
+  auto type = variable::variableFactory().elem_dtype(data);
+  // Convert everything other than float to dtype<double>.
+  type = type == dtype<float> ? dtype<float> : dtype<double>;
+
+  if (data.dtype() == dtype<bucket<DataArray>>) {
+    const auto &&[indices, dim, buffer] = data.constituents<DataArray>();
+    if (const auto mask_union = irreducible_mask(buffer.masks(), dim);
+        mask_union.is_valid()) {
+      // Trick to get the sizes of bins if masks are present - bin the masks
+      // using the same dimension & indices as the data, and then sum the
+      // inverse of the mask to get the number of unmasked entries.
+      return astype(buckets::sum(data), type, CopyPolicy::TryAvoid) /
+             astype(
+                 buckets::sum(make_bins_no_validate(indices, dim, ~mask_union)),
+                 type, CopyPolicy::TryAvoid);
+
+    }
+  }
+  return astype(buckets::sum(data), type, CopyPolicy::TryAvoid) /
+         astype(bucket_sizes(data), type, CopyPolicy::TryAvoid);
```

Comment:
I just had a look at the implementation of `mean` (for non-binned data) and notice that there is a helper, `normalize_impl` in `variable/operations_common.h` which can do much of the work in the code above, I think you should use it.